### PR TITLE
NN-1918 IEP Role Access

### PR DESCRIPTION
**Pull request by bell-pa**
New endpoint for assigning  roles to, and removing roles from users who match given criteria.
Criteria are membership of named caseloads and roles.

### CODE REVIEW

File: src/main/java/uk/gov/justice/digital/hmpps/keyworker/controllers/RoleMangementController.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import uk.gov.justice.digital.hmpps.keyworker.dto.RoleAssignmentsSpecification;
+import uk.gov.justice.digital.hmpps.keyworker.rolemigration.RoleAssignmentsService;
+
+@Api(tags = {"caseloads-roles"})
+
+@RestController
+@RequestMapping(value = "caseloads-roles")
+@Slf4j
+public class RoleMangementController {
+
```
Comment (bell-pa) [inline src/main/java/uk/gov/justice/digital/hmpps/keyworker/controllers/RoleMangementController.java]: Spelling!

File: src/main/java/uk/gov/justice/digital/hmpps/keyworker/rolemigration/RoleAssignmentsService.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        this.roleService = roleService;
+    }
+
+    /**
+     * Look up users by caseload (prison) and membership of specification.rolesToMatch.
+     * for each matched user:
+     * Assign to the user the roles identified in specification.rolesToAssign.
+     * Remove from the user (at the caseload) the roles identified by specification.rolesToRemove.
+     */
+    @PreAuthorize("hasAnyRole('MAINTAIN_ACCESS_ROLES', 'MAINTAIN_ACESS_ROLES_ADMIN')")
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/digital/hmpps/keyworker/rolemigration/RoleAssignmentsService.java]: `MAINTAIN_ACESS_ROLES_ADMIN` should be `ACCESS`?

File: src/main/java/uk/gov/justice/digital/hmpps/keyworker/rolemigration/RoleAssignmentsService.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+            val usernamesForCaseload = findUsernamesMatchingRolesAtCaseload(specification.getRolesToMatch(), caseload);
+            log.debug("Found {} users for the {} caseload: {}.", usernamesForCaseload.size(), caseload, usernamesForCaseload);
+            usernamesForCaseload.forEach(username -> {
+                try {
+                    assignRolesToUser(username, specification.getRolesToAssign());
+                    try {
+                        removeRolesFromUserAtCaseload(caseload, username, specification.getRolesToRemove());
+                    } catch (Exception e) {
+                        log.debug("Failure while removing roles {} from user {} at caseload {}. Some roles may not have been removed.", specification.getRolesToRemove(), username, caseload);
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/digital/hmpps/keyworker/rolemigration/RoleAssignmentsService.java]: bit naff catching `Exception` and then not logging the exception in any way.  What could cause the exception to be thrown?  Why is this `debug` - isn't this an `error` in our code somehow?

File: src/main/java/uk/gov/justice/digital/hmpps/keyworker/controllers/RoleManagementController.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+            nickname = "assignRolesFormUrlEncoded",
+            consumes = MediaType.APPLICATION_FORM_URLENCODED_VALUE
+    )
+    @ApiResponses({
+            @ApiResponse(code = 204, message = "Successful"),
+            @ApiResponse(code = 400, message = "Client error")
+    })
+    @PostMapping(consumes = MediaType.APPLICATION_FORM_URLENCODED_VALUE)
+    public ResponseEntity assignRolesForm(@ApiParam(value = "An application/x-www-form-urlencoded form.  Keys 'caseloads' and 'rolesToMatch' are mandatory, 'rolesToAdd' and 'rolesToRemove' should be supplied as needed") @RequestParam MultiValueMap<String, String> form) {
+        roleAssignmentsService.updateRoleAssignments(RoleAssignmentsSpecification.fromForm(form));
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/digital/hmpps/keyworker/controllers/RoleManagementController.java]: could this call `assignRolesJson` instead, which removes code duplication

File: src/main/java/uk/gov/justice/digital/hmpps/keyworker/rolemigration/RoleAssignmentsService.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    @PreAuthorize("hasAnyRole('MAINTAIN_ACCESS_ROLES', 'MAINTAIN_ACESS_ROLES_ADMIN')")
+    public void updateRoleAssignments(@Valid final RoleAssignmentsSpecification specification) {
+        log.debug("Updating role assignments: {}", specification);
+        for (val caseload : specification.getCaseloads()) {
+
+            val usernamesForCaseload = findUsernamesMatchingRolesAtCaseload(specification.getRolesToMatch(), caseload);
+            log.debug("Found {} users for the {} caseload: {}.", usernamesForCaseload.size(), caseload, usernamesForCaseload);
+            usernamesForCaseload.forEach(username -> {
+                try {
+                    assignRolesToUser(username, specification.getRolesToAssign());
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/digital/hmpps/keyworker/rolemigration/RoleAssignmentsService.java]: what will happen if the user already has the role - i.e. the code is re-run?

File: src/test/groovy/uk/gov/justice/digital/hmpps/keyworker/integration/specs/RoleAssignmentSpecification.groovy [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+import org.springframework.http.HttpHeaders
+import org.springframework.http.HttpMethod
+import org.springframework.http.HttpStatus
+import org.springframework.http.MediaType
+import uk.gov.justice.digital.hmpps.keyworker.dto.RoleAssignmentsSpecification
+
+class RoleAssignmentSpecification extends TestSpecification {
+
+    def "A user that has the MAINTAIN_ACCESS_ROLES_ADMIN role can make role assignment changes"() {
+        given: "That I have the MAINTAIN_ACESS_ROLES_ADMIN role"
```
Comment (petergphillips) [inline src/test/groovy/uk/gov/justice/digital/hmpps/keyworker/integration/specs/RoleAssignmentSpecification.groovy]: should be `ACCESS`

File: src/test/groovy/uk/gov/justice/digital/hmpps/keyworker/integration/specs/RoleAssignmentSpecification.groovy [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+        HttpHeaders headers = new HttpHeaders()
+        headers.setBearerAuth(adminToken)
```
Comment (petergphillips) [inline src/test/groovy/uk/gov/justice/digital/hmpps/keyworker/integration/specs/RoleAssignmentSpecification.groovy]: the `adminToken` doesn't have the `MAINTAIN_ACCESS_ROLES_ADMIN` role though :confused:, it has the `ROLE_MAINTAIN_ACCESS_ROLES` role

File: src/main/java/uk/gov/justice/digital/hmpps/keyworker/rolemigration/RoleAssignmentsService.java (hunk 2 above)
Comment (Mjwillis) [reply src/main/java/uk/gov/justice/digital/hmpps/keyworker/rolemigration/RoleAssignmentsService.java]: Don't think MAINTAIN_ACCESS_ROLES should be allowed

File: src/test/groovy/uk/gov/justice/digital/hmpps/keyworker/integration/specs/TestSpecification.groovy [hunk 8]
```diff
@@ -27,8 +27,8 @@ import static org.springframework.boot.test.context.SpringBootTest.WebEnvironmen
...
+    protected String token = "eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJpbnRlcm5hbFVzZXIiOnRydWUsInVzZXJfbmFtZSI6IklUQUdfVVNFUiIsInNjb3BlIjpbInJlYWQiLCJ3cml0ZSJdLCJleHAiOjE4NDc4ODY4MTQsImF1dGhvcml0aWVzIjpbIlJPTEVfS1dfQURNSU4iLCJST0xFX09NSUNfQURNSU4iXSwianRpIjoiMzA1MjcwZjItYzZiYS00ZWY5LWIxOGMtYTllZTE1Zjk0ZTE2IiwiY2xpZW50X2lkIjoiZWxpdGUyYXBpY2xpZW50In0.WbqnPlPqr08tC7PEilLOrkzfh0UziyuD9lFnTUVbCQIK8Gp0c4iP9aJ0NydiPhcRzoUfaA9eUoZyMBp1GwpBa5kn-YNax4mBfL0pXT-zl6c1Bx6VqmVCNmAT5BXl2tkfAi0DnbRHHAqwaAR4UlGe5JSMt6mloomGqNFHkAuJinndR1LC9zlm0473LnLgmDaCVEuJDlUbaL660o5nHSEimC2yNpm8VI7_i5nYSzECpKw_BYkBuiuIQJp_9-BLcf67Q3FkxM1NDqQIPwfKtTQ2ayJJEWdaI0Z8PrwxTlDFwm7nfSBSf-rXSXaHaL_9BBnsdboOx5lEpl5p7f1A2bOjag"
```
Comment (Mjwillis) [inline src/test/groovy/uk/gov/justice/digital/hmpps/keyworker/integration/specs/TestSpecification.groovy]: what has changed in token?

File: src/test/groovy/uk/gov/justice/digital/hmpps/keyworker/integration/specs/RoleAssignmentSpecification.groovy (hunk 7 above)
Comment (Mjwillis) [reply src/test/groovy/uk/gov/justice/digital/hmpps/keyworker/integration/specs/RoleAssignmentSpecification.groovy]: again that role should not allow access - remove ROLE_MAINTAIN_ACCESS_ROLES only the admin one

File: src/main/java/uk/gov/justice/digital/hmpps/keyworker/rolemigration/RoleAssignmentsService.java (hunk 5 above)
Comment (Mjwillis) [reply src/main/java/uk/gov/justice/digital/hmpps/keyworker/rolemigration/RoleAssignmentsService.java]: Its also worth considering the restartability of this API.  ie. if one nomis call fails, do we stop, retry, etc.

File: src/main/java/uk/gov/justice/digital/hmpps/keyworker/rolemigration/RoleAssignmentsService.java (hunk 3 above)
Comment (bell-pa) [reply src/main/java/uk/gov/justice/digital/hmpps/keyworker/rolemigration/RoleAssignmentsService.java]: Changed to log.warn(message, exception).

The exceptions are likely to be due to remote calls failing, not errors in the code.

File: src/main/java/uk/gov/justice/digital/hmpps/keyworker/controllers/RoleManagementController.java (hunk 4 above)
Comment (bell-pa) [reply src/main/java/uk/gov/justice/digital/hmpps/keyworker/controllers/RoleManagementController.java]: Wouldn't that be a bit weird?

File: src/main/java/uk/gov/justice/digital/hmpps/keyworker/rolemigration/RoleAssignmentsService.java (hunk 5 above)
Comment (bell-pa) [reply src/main/java/uk/gov/justice/digital/hmpps/keyworker/rolemigration/RoleAssignmentsService.java]: The end-point works by caseload, then by matched user within caseload.
The end-point calls elite2api end-points to add and remove roles from users (at caseloads).
The remote calls are repeatable; you can add the same role as many times as you like. Removing a role from a user who does not have the role results in a 404.
The end-point swallows the 404 errors and continues. 
More serious errors stop processing for the current user at the current caseload. Work resumes with the next user/caseload combination.  Adds are done before removes - so all roles should be added to a user before any roles are removed from that user.
Therefore the end-point can be used repeatedly with the same payload, and eventually, (on first call if the remote calls are all OK) the state of the role assignments should settle to what is specified in the payload.
I've tried various combinations of add/remove in different sequences against T3.  It seems to work reasonably well.
The logs should give plenty of information about what happened.
I can demo on Tuesday if that helps...

File: src/test/groovy/uk/gov/justice/digital/hmpps/keyworker/integration/specs/TestSpecification.groovy (hunk 8 above)
Comment (bell-pa) [reply src/test/groovy/uk/gov/justice/digital/hmpps/keyworker/integration/specs/TestSpecification.groovy]: I changed the access from private to protected so I could access the token in my tests.
Now I've reverted the change and created new tokens to use with the integration tests.

**Comment by bell-pa** [comment]
1) Adding app insight tooling could be covered by a separate story.  
2) The api could return information about what it did. - summary or detail.
3) Could be re-worked to use Spring Batch or similar if needed.

**Pull request merged**